torch/_higher_order_ops/cond.py

```diff
@@ -28,11 +30,51 @@
 )
 
 
+@contextmanager
+def _set_compilation_env():
+    _old_is_tracing = torch.fx._symbolic_trace._is_fx_tracing_flag
+    _old_capture_dyn_ops = torch._dynamo.config.capture_dynamic_output_shape_ops
+    _old_capture_scalar = torch._dynamo.config.capture_scalar_outputs
```

Comment:
For is_fx_tracing_flag, dynamo uses it to check whether we're in a tracin session [here](https://github.com/pytorch/pytorch/blob/71045f4885c1d8e684a0e4f2b64e2edf348c541a/torch/_dynamo/eval_frame.py#L311C29-L311C30). Right now, we have no problem in tracing make_fx( mode="real") so don't want to raise error on it since that's a lot of expectedFailures. We could consider remove the check in dynamo when things are ready.

For capture_dynamic_output_shape_ops and capture_scalar_outputs, they're used fro operations such as .nonzero and .item, we'd like to trace through them instead of having a graph break which will violate the fullgraph=True and a test failure.

Reply:
Can you point me to a test (or some sample code) where the `is_fx_tracing_flag` influences?

> For capture_dynamic_output_shape_ops and capture_scalar_outputs, they're used fro operations such as .nonzero and .item, we'd like to trace through them instead of having a graph break which will violate the fullgraph=True and a test failure.

Are there specific tests that you're looking at? Like I would have expected these to not work with regular torch.compile/export, and if so, these should error out with eager-mode cond.

Reply:
> where the is_fx_tracing_flag influences

Yeah, basically all make_fx calls (despite the mode) will switch this flag to True. make_fx -> dispatch_trace -> tracer_base.trace -> https://github.com/pytorch/pytorch/blob/main/torch/fx/_symbolic_trace.py#L726 and have many tests in test_control_flow.py uses make_fx.

> Are there specific tests that you're looking at?

I couldn't find a concrete example but maybe we should try a bit on CI? It seems to me quite restrictive for cond not be able to call .item or nonzero inside. [export](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L1126) actually has a config to turn them on by default.